agent/xds/routes.go

```diff
@@ -24,28 +25,103 @@ func routesFromSnapshot(cInfo connectionInfo, cfgSnap *proxycfg.ConfigSnapshot)
 
 	switch cfgSnap.Kind {
 	case structs.ServiceKindConnectProxy:
-		return routesFromSnapshotConnectProxy(cInfo, cfgSnap)
+		return routesForConnectProxy(cInfo, cfgSnap.Proxy.Upstreams, cfgSnap.ConnectProxy.DiscoveryChain)
 	case structs.ServiceKindIngressGateway:
-		return routesFromSnapshotIngressGateway(cInfo, cfgSnap)
+		return routesForIngressGateway(cInfo, cfgSnap.IngressGateway.Upstreams, cfgSnap.IngressGateway.DiscoveryChain)
+	case structs.ServiceKindTerminatingGateway:
+		return routesFromSnapshotTerminatingGateway(cInfo, cfgSnap)
 	default:
 		return nil, fmt.Errorf("Invalid service kind: %v", cfgSnap.Kind)
 	}
 }
 
-// routesFromSnapshotConnectProxy returns the xDS API representation of the
-// "routes" in the snapshot.
-func routesFromSnapshotConnectProxy(cInfo connectionInfo, cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
+// routesFromSnapshotTerminatingGateway returns the xDS API representation of the "routes" in the snapshot.
+// For any HTTP service we will return a default route.
+func routesFromSnapshotTerminatingGateway(_ connectionInfo, cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
 	if cfgSnap == nil {
 		return nil, errors.New("nil config given")
 	}
 
 	var resources []proto.Message
-	for _, u := range cfgSnap.Proxy.Upstreams {
+	for svc := range cfgSnap.TerminatingGateway.ServiceGroups {
```

Comment:
Traverse the `TerminatingGateway.ValidServices()` slice instead. That'll take care of ensuring all of the watches have data for each service.

Reply:
Fixed